csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +519,191 @@ class RedundantUseAnalysis : BackwardVisitor {
 
 } // namespace
 
+// This function removes the redundant write to gmem when an output tensor has a
+// leaf domain merged from concretized broadcast root domain and parallelized by
+// thread/block id. issue https://github.com/csarofeen/pytorch/issues/2125
+void ThreadPredicateMap::avoidConcretizedBroadcastRedundantWrite(
+    const TensorView* out_tv) {
+  // step-1, check if there is a leaf domain parallelized by thread/block id and
+  // its definition is a merged
+  bool merged_parallelized_thread_block = false;
+  for (auto ld : out_tv->domain()->domain()) {
+    const ParallelType& pt = ld->getParallelType();
+    auto merge = dynamic_cast<Merge*>(ld->definition());
+    if (isParallelTypeThread(pt) && merge) {
+      merged_parallelized_thread_block = true;
+      break;
+    }
+  }
+  // shortcut if there is no such leaf domain
+  if (!merged_parallelized_thread_block) {
+    return;
+  }
+
+  auto root_domain = out_tv->getRootDomain();
+  // For each broadcast root domain, find a concretized domain from its
+  // exact mapped domain set.
+  auto getConcretizedBroadcastRootDomain = [&]() {
+    std::unordered_map<IterDomain*, IterDomain*>
+        concretized_broadcast_root_domains;
+    for (auto rd : root_domain) {
+      if (rd->isBroadcast()) {
+        const auto& mapped_rd_exact_set = GpuLower::current()
+                                              ->caMap()
+                                              ->getIdSets(IdMappingMode::EXACT)
+                                              .getDisjointSetOf(rd);
+        for (auto mapped_rd : mapped_rd_exact_set) {
+          std::unordered_set<IterDomain*> all_cids =
+              GpuLower::current()
+                  ->concretizedBroadcastDomains()
+                  ->allConcretizedDomains(mapped_rd);
+          if (!all_cids.empty()) {
+            concretized_broadcast_root_domains[rd] = *all_cids.begin();
+            break;
+          }
+        }
+      }
+    }
+    return concretized_broadcast_root_domains;
+  };
+
+  // Find all the root domains that are merged to the leaf domain.
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}
+  auto getRootDomainsMergedToLeaf = [&](IterDomain* ld) {
+    std::vector<IterDomain*> merged_root_domains;
+    std::vector<int> index_root_domain;
+    auto all_exp = DependencyCheck::getAllExprsBetween(
+        {root_domain.begin(), root_domain.end()}, {ld});
+    for (auto expr : all_exp) {
+      if (auto merge = dynamic_cast<Merge*>(expr)) {
+        auto outer_iter =
+            std::find(root_domain.begin(), root_domain.end(), merge->outer());
+        auto inner_iter =
+            std::find(root_domain.begin(), root_domain.end(), merge->inner());
+        // outer_iter and inner_iter should be neighbors in root_domain
+        if (inner_iter != root_domain.end() &&
+            outer_iter != root_domain.end() &&
+            std::distance(outer_iter, inner_iter) != 1) {
```

Comment:
The if-statement is trying to stop applying this thread/block predication if there is a merge where the merged root domains are not neighbors, e.g. [I1, B2, B3] with merges (I1*B3) will be rejected because `I1` is a root domain, `B3` is a root domain, and their distance is not 1.
It won't influence cases where one of the merged domain is not a root domain. e.g. in the case you mensioned, `I0, I1, I2 and merge(merge(I0, I1), I2)`, the if-statement test is false. so the case will not be rejcted. and the predication will be added for this case.

Your suggested approach is more general and I implemented it in the reivsed version.